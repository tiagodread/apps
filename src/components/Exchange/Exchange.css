/* Align exchange-container to the center of the screen */
.exchange-container {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}