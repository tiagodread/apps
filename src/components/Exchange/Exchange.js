import { React, useState } from "react";
import {
  FormControl,
  InputLabel,
  MenuItem,
  Select,
  Input,
  InputAdornment,
  Alert,
  Button,
} from "@mui/material";
import { Container } from "@mui/system";
import PropTypes from "prop-types";
import "./Exchange.css";

function Exchange(props) {
  const [fromCurrency, setFromCurrency] = useState(props.fromCurrency);
  const [toCurrency, setToCurrency] = useState(props.toCurrency);
  const [value, setValue] = useState(undefined);
  const [convertedValue, setConvertedValue] = useState("");
  const [exchangeRate, setExchangeRate] = useState("");
  const [showResults, setShowResults] = useState(false);
  const [showErrors, setShowErrors] = useState(false);

  const handleChangeFrom = (event) => {
    setFromCurrency(event.target.value);
  };

  const handleChangeTo = (event) => {
    setToCurrency(event.target.value);
  };

  const handleExchangeRate = (value) => {
    setExchangeRate(value);
  };

  const handleConvertedValue = (value) => {
    setConvertedValue(value);
  };

  const handleValue = (event) => {
    setShowResults(false);
    setShowErrors(false);
    setValue(event.target.value);
  };

  const getExchangeRate = () => {
    const apiKey = "7c478081950780fb862c";
    let query = `${fromCurrency}_${toCurrency}`;
    let url = `https://free.currconv.com/api/v7/convert?q=${query}&compact=ultra&apiKey=${apiKey}`;
    fetch(url)
      .then((response) => response.json())
      .then((data) => {
        const exchangeRate = parseFloat(data[query]).toFixed(4);
        handleExchangeRate(exchangeRate);
        const convertedValue = parseFloat(value * exchangeRate).toFixed(2);
        handleConvertedValue(convertedValue);
      });
  };

  const convert = () => {
    if (value === undefined || value < 1) {
      setShowErrors(true);
      return;
    }
    if (fromCurrency === toCurrency) {
      setConvertedValue(value);
      return;
    }
    getExchangeRate();
    setShowResults(true);
  };

  return (
    <Container className="exchange-container">
      <h1>ðŸ’± Currency Converter</h1>
      <form>
        <div>
          <FormControl fullWidth>
            <InputLabel id="exchange-form-from">From Currency</InputLabel>
            <Select
              labelId="exchange-form-from"
              id="exchange-form-from"
              data-testid="from-currency"
              value={fromCurrency}
              label="From Currency"
              required={true}
              onChange={handleChangeFrom}
            >
              <MenuItem value="EUR">EUR</MenuItem>
              <MenuItem value="USD">USD</MenuItem>
              <MenuItem value="GBP">GBP</MenuItem>
              <MenuItem value="JPY">JPY</MenuItem>
              <MenuItem value="BRL">BRL</MenuItem>
              <MenuItem value="CAD">CAD</MenuItem>
            </Select>
            <br />
          </FormControl>

          <FormControl fullWidth>
            <InputLabel id="exchange-form-to">To Currency</InputLabel>
            <Select
              labelId="exchange-form-to"
              id="exchange-form-to"
              data-testid="to-currency"
              value={toCurrency}
              label="To Currency"
              onChange={handleChangeTo}
              required={true}
            >
              <MenuItem value="EUR">EUR</MenuItem>
              <MenuItem value="USD">USD</MenuItem>
              <MenuItem value="GBP">GBP</MenuItem>
              <MenuItem value="JPY">JPY</MenuItem>
              <MenuItem value="BRL">BRL</MenuItem>
              <MenuItem value="CAD">CAD</MenuItem>
            </Select>
            <br />
          </FormControl>

          <FormControl fullWidth sx={{ m: 1 }} variant="outlined">
            <InputLabel htmlFor="amount">Amount</InputLabel>
            <Input
              type="number"
              id="amount"
              data-testid="enteredValue"
              placeholder={`value in ${fromCurrency} format`}
              onChange={(event) => handleValue(event)}
              required={true}
              startAdornment={
                <InputAdornment position="start">$</InputAdornment>
              }
            />
          </FormControl>
        </div>
        <Button
          variant="primary"
          data-testid="convert"
          value="convert"
          onClick={convert}
        >
          Convert
        </Button>{" "}
      </form>

      {showResults ? (
        <Alert
          variant="outlined"
          severity="success"
          className="exchange-results"
          data-testid="exchange-result"
        >
          <p>
            {value} {fromCurrency} = {convertedValue} {toCurrency}
          </p>
          <hr />
          <p className="mb-0">Current exchange Rate: {exchangeRate}</p>
        </Alert>
      ) : null}

      {showErrors ? (
        <Alert
          variant="outlined"
          severity="error"
          className="exchange-results"
          data-testid="exchange-result"
        >
          <p>Please enter a valid value!</p>
        </Alert>
      ) : null}
    </Container>
  );
}

Exchange.propTypes = {
  fromCurrency: PropTypes.string,
  toCurrency: PropTypes.string,
};

Exchange.defaultProps = {
  fromCurrency: "BRL",
  toCurrency: "USD",
};

export default Exchange;
