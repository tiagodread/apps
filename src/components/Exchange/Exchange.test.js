import { render, screen, fireEvent } from "@testing-library/react";
import Exchange from "./Exchange";

describe("Exchange", () => {
  it("renders without crashing", () => {
    render(<Exchange />);
  });

  it("Should match the snapshot", () => {
    const { container } = render(<Exchange />);
    expect(container).toMatchSnapshot();
  });

  it("renders the correct initial currency state", () => {
    render(<Exchange fromCurrency="CAD" toCurrency="EUR" />);
    const fromCurrency = screen.getByTestId("from-currency");
    expect(fromCurrency.value).toBe("CAD");
    const toCurrency = screen.getByTestId("to-currency");
    expect(toCurrency.value).toBe("EUR");
  });

  it("should change currency state when the user changes the from currency", () => {
    render(<Exchange fromCurrency="CAD" toCurrency="EUR" />);
    const fromCurrency = screen.getByTestId("from-currency");
    expect(fromCurrency.value).toBe("CAD");
    fromCurrency.value = "USD";
    expect(fromCurrency.value).toBe("USD");
  });

  it("should change currency state when the user changes the to currency", () => {
    render(<Exchange fromCurrency="CAD" toCurrency="EUR" />);
    const toCurrency = screen.getByTestId("to-currency");
    expect(toCurrency.value).toBe("EUR");
    toCurrency.value = "USD";
    expect(toCurrency.value).toBe("USD");
  });

  it("should show an window alert if the value is undefined or 0", () => {
    jest.spyOn(window, "alert").mockImplementation(() => {});
    render(<Exchange fromCurrency="CAD" toCurrency="EUR" />);
    const convertButton = screen.getByTestId("convert");
    fireEvent.click(convertButton);
    expect(window.alert).toHaveBeenCalled();
    const enteredValue = screen.getByTestId("enteredValue");
    enteredValue.value = "0";
    fireEvent.click(convertButton);
    expect(window.alert).toHaveBeenCalled();
  });

  it("Should fetch the exchange rate when the user clicks the convert button", async () => {
    render(<Exchange fromCurrency="CAD" toCurrency="EUR" />);
    const enteredValue = screen.getByTestId("enteredValue");
    fireEvent.change(enteredValue, { target: { value: "5" } });
    const convertButton = screen.getByTestId("convert");
    fireEvent.click(convertButton);
    await screen.findByTestId("exchange-result");
  });
});
