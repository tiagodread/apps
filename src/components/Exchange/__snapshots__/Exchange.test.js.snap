// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Should match the snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg exchange-container css-1oqqzyl-MuiContainer-root"
  >
    <h1>
      ðŸ’± Currency Converter
    </h1>
    <form>
      <div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="exchange-form-from"
          >
            From Currency
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            data-testid="from-currency"
            required=""
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="exchange-form-from exchange-form-from"
              class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="exchange-form-from"
              role="button"
              tabindex="0"
            >
              BRL
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              required=""
              tabindex="-1"
              value="BRL"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1in441m"
              >
                <span>
                  From Currency
                  Â 
                  *
                </span>
              </legend>
            </fieldset>
          </div>
          <br />
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="exchange-form-to"
          >
            To Currency
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            data-testid="to-currency"
            required=""
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="exchange-form-to exchange-form-to"
              class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="exchange-form-to"
              role="button"
              tabindex="0"
            >
              USD
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              required=""
              tabindex="-1"
              value="USD"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1in441m"
              >
                <span>
                  To Currency
                  Â 
                  *
                </span>
              </legend>
            </fieldset>
          </div>
          <br />
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-1w4vsez-MuiFormControl-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="amount"
          >
            Amount
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1ptx2yq-MuiInputBase-root-MuiInput-root"
            data-testid="enteredValue"
            required=""
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
              >
                $
              </p>
            </div>
            <input
              aria-invalid="false"
              class="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedStart css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="amount"
              placeholder="value in BRL format"
              required=""
              type="number"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButtonBase-root  css-1w1rijm-MuiButtonBase-root-MuiButton-root"
        data-testid="convert"
        tabindex="0"
        type="button"
        value="convert"
      >
        Convert
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
       
    </form>
  </div>
</div>
`;
